File: object-oriented-programming/virtual-world-java/src/Game/Model/Organism.java
package Game.Model;

import java.awt.*;
import java.io.Serializable;

public abstract class Organism implements Serializable {

    protected Species species;
    protected int y;
    protected int x;
    protected int strength;
    protected int initiative;
    protected int age = 0;
    protected boolean dead = false;
    protected World world;
    protected int breedCooldown;
    protected boolean canBreed;

    public Organism(Species species, int x, int y, int strength, int initiative, World world, int breedCooldown, boolean canBreed){
        this.species = species;
        this.y = y;
        this.x = x;
        this.strength = strength;
        this.initiative = initiative;
        this.world = world;
        this.breedCooldown = breedCooldown;
        this.canBreed = canBreed;

        world.setOrganism(this, y, x);
        world.addOrganism(this);
    }

    public abstract void action();
    public abstract void collision(Organism other);
    public abstract Color draw();

    public void moveOrganism(int newY, int newX){
        Organism occupant = world.getOrganism(newY, newX);
        if (occupant != null) {
            occupant.collision(this);
        } else {
            world.setOrganism(null, y, x);
            y = newY;
            x = newX;
            world.setOrganism(this, y, x);
        }
    }

    public final int getY() { return y; }
    public final int getX() { return x; }
    public final int getStrength() { return strength; }
    public final int getInitiative() { return initiative; }
    public final int getAge() { return age; }
    public final Species getSpecies() { return species; }
    public void incrementAge() { age++; }
    public void incrementStrength(int number) { strength+=number; }
    public boolean isDead() { return dead; }
    public void markDead() { dead = true; }
    public boolean checkCanBreed() { return canBreed; }
    public void setBreedCooldown() {
        breedCooldown = 3;
        canBreed = false;
    }
}

File: object-oriented-programming/virtual-world-python/organism.py
from abc import abstractmethod

class Organism :
    
    def __init__(self, species, x, y, strength, initiative, world, can_breed, breed_cooldown) :
        self._species = species
        self._x = x
        self._y = y
        self._age = 0
        self._dead = False
        self._strength = strength
        self._initiative = initiative
        self._world = world
        self._can_breed = can_breed
        self._breed_cooldown = breed_cooldown

        world.set_organism(self, x, y)
        world.add_organism(self)

    @abstractmethod
    def action(self) :
        pass

    @abstractmethod
    def collision(self, other) :
        pass

    @abstractmethod
    def draw(self) :
        pass

    def move_organism(self, new_x, new_y) :
        occupant = self._world.get_organism(new_x, new_y)
        if (occupant) :
            occupant.collision(self)
        else :
            self._world.set_organism(None, self.x, self.y)
            self._x = new_x
            self._y = new_y
            self._world.set_organism(self, self.x, self.y)

    @property
    def y(self) :
        return self._y

    @property
    def x(self) :
        return self._x
    
    @property
    def strength(self) :
        return self._strength
    
    @property
    def initiative(self) :
        return self._initiative
    
    @property
    def age(self) :
        return self._age
    
    @property
    def species(self) :
        return self._species
    
    def increment_age(self) :
        self._age+=1

    def increment_strength(self, number) :
        self._strength += number

    def is_dead(self) :
        return self._dead
    
    def mark_dead(self) :
        self._dead = True

    def check_can_breed(self) :
        return self._can_breed
    
    def set_breed_cooldown(self) :
        self._breed_cooldown = 3
        self._can_breed = False
        

